monai/handlers/checkpoint_saver.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from ignite.engine import Events
+from ignite.handlers import ModelCheckpoint
+
+
+class CheckpointSaver:
+    """
+    CheckpointSaver acts as an Ignite handler to save checkpoint data into files.
+    It supports to save according to metrics result, epoch number, iteration number
+    and last model or exception.
+
+    Args:
+        save_dir (str): the target directory to save the checkpoints.
+        save_dict (dict): source objects that save to the checkpoint. examples::
+
+            {'network': net, 'optimizer': optimizer, 'engine', engine}
+
+        name (str): identifier of logging.logger to use, if None, defaulting to ``engine.logger``.
+        file_prefix (str): prefix for the filenames to which objects will be saved.
+        save_final (bool): whether to save checkpoint or session at final iteration or exception.
+        save_key_metric (bool): whether to save checkpoint or session according to key metric value.
+        key_metric_name (str): the name of key_metric in ignite metrics dictionary.
+            if None, use `engine.state.key_metric` instead.
+        key_metric_n_saved (int): save top N checkpoints or sessions sorted by the key metric.
+        epoch_save_interval (int): save checkpoint or session every N epochs.
+        epoch_n_saved (int): save latest N epochs' checkpoints or sessions, 'None' is to save all.
+        iteration_save_interval (int): save checkpoint or session every N iterations.
+        iteration_n_saved (int): save latest N iterations' checkpoints or sessions, 'None' is to save all.
+        convert_to_single_gpu (bool): if current model is trained by multi GPU and
```

Comment:
do we need an option for convert_to_multi_gpu? Iooks like it should always be True, and it won't break the existing single gpu functionality

Reply:
The problem is that:
1. There are many items in "save_dict" but only the network can change to multi_gpu mode, need to set another parameter to indicate the network object.
2. It's easy to remove `module` in network but not easy to add.

What do you think?

Reply:
looks like pytorch recommends remove `module` https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-torch-nn-dataparallel-models

Reply:
Yes, as the PyTorch doc said: "To save a DataParallel model generically, save the model.module.state_dict()."
That's why I added the option `convert_to_single_gpu ` but no `convert_to_multi_gpu `.
Do you still want to add `convert_to_multi_gpu`?

Reply:
sorry, I made the wrong comment, I was thinking, could we remove `convert_to_single_gpu` option?

Reply:
Actually, that's what I did in Clara before, but in order to support more flexibility, I added this option...I will remove `module` directly in next patch.
Thanks.

Reply:
updated in the latest patch.
Thanks.